File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/NotificationsViewAdapter.java
package com.amazon.tv.leanbacklauncher.notifications;

import android.content.Context;
import android.os.Handler;
import android.util.Log;

import androidx.collection.ArraySet;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.DiffUtil.Callback;
import androidx.recyclerview.widget.DiffUtil.DiffResult;
import androidx.recyclerview.widget.ListUpdateCallback;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import com.amazon.tv.leanbacklauncher.MainActivity;
import com.amazon.tv.leanbacklauncher.util.Preconditions;
import com.amazon.tv.leanbacklauncher.widget.RowViewAdapter;
import com.amazon.tv.tvrecommendations.TvRecommendation;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public abstract class NotificationsViewAdapter<VH extends ViewHolder> extends RowViewAdapter<VH> implements MainActivity.IdleListener {
    static final int UI_STATE_VISIBLE = 1;
    private HomeScreenMessaging mHomeScreenMessaging;
    private final ArrayList<TvRecommendation> mMasterList = new ArrayList<>();
    private PrioritizeRowUpdateState<VH> mPrioritizeRowUpdateState;
    private final Set<TvRecommendation> mRecommendationToBeRemoved = new ArraySet();
    private final ArrayList<TvRecommendation> mSyncedList = new ArrayList();
    private int mUiState = 0;

    static class PrioritizeRowUpdateState<VH extends ViewHolder> {
        private final NotificationsViewAdapter<VH> mAdapter;
        private final Handler mHandler = new Handler();
        private boolean mIsIdle = false;
        private final long mMaxUpdateTime;
        private final long mMinUpdateTime;
        private final Runnable mStateTickRunnable = new Runnable() {
            public void run() {
                PrioritizeRowUpdateState.this.updateTick();
            }
        };

        PrioritizeRowUpdateState(NotificationsViewAdapter<VH> adapter, long minUpdateTime, long maxUpdateTime) {
            this.mAdapter = adapter;
            this.mMinUpdateTime = minUpdateTime;
            this.mMaxUpdateTime = maxUpdateTime;
        }

        private void scheduleUpdateTickIfNeeded() {
            List<TvRecommendation> masterList = this.mAdapter.getMasterList();
            List<TvRecommendation> syncedList = this.mAdapter.getSyncedList();
            this.mHandler.removeCallbacks(this.mStateTickRunnable);
            if (masterList.size() - syncedList.size() > 0) {
                this.mHandler.postDelayed(this.mStateTickRunnable, ((this.mMaxUpdateTime - this.mMinUpdateTime) / ((long) (masterList.size() - syncedList.size()))) + this.mMinUpdateTime);
            }
        }

        private void unScheduleUpdateTick() {
            this.mHandler.removeCallbacks(this.mStateTickRunnable);
        }

        private void postDeletesAndSubstitutes() {
            final List<TvRecommendation> masterList = this.mAdapter.getMasterList();
            final List<TvRecommendation> syncedList = this.mAdapter.getSyncedList();
            DiffResult result = DiffUtil.calculateDiff(new Callback() {
                public int getOldListSize() {
                    return syncedList.size();
                }

                public int getNewListSize() {
                    return masterList.size();
                }

                public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                    return NotificationUtils.equals(syncedList.get(oldItemPosition), masterList.get(newItemPosition));
                }

                public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                    return Objects.equals(syncedList.get(oldItemPosition), masterList.get(newItemPosition));
                }

                public Object getChangePayload(int oldItemPosition, int newItemPosition) {
                    return masterList.get(newItemPosition);
                }
            }, false);
            List<TvRecommendation> snapshot = new ArrayList(syncedList);
            result.dispatchUpdatesTo(new ListUpdateCallback() {
                public void onInserted(int position, int count) {
                    for (int i = 0; i < count; i++) {
                        syncedList.add(position + i, null);
                    }
                }

                public void onRemoved(int position, int count) {
                    for (int i = 0; i < count; i++) {
                        syncedList.remove(position);
                    }
                }

                public void onMoved(int fromPosition, int toPosition) {
                    throw new IllegalStateException("Should not receive move events");
                }

                public void onChanged(int position, int count, Object payload) {
                    for (int i = 0; i < count; i++) {
                        syncedList.set(position + i, (TvRecommendation) payload);
                    }
                }
            });
            Iterator<TvRecommendation> iterator = syncedList.iterator();
            while (iterator.hasNext()) {
                if (iterator.next() == null) {
                    iterator.remove();
                }
            }
            this.mAdapter.notifyChangesSinceSnapshot(snapshot);
            this.mAdapter.updateRowVisibility();
        }

        private void postOneInsert() {
            List<TvRecommendation> masterList = this.mAdapter.getMasterList();
            List<TvRecommendation> syncedList = this.mAdapter.getSyncedList();
            int i = 0;
            while (i < masterList.size() && i <= syncedList.size()) {
                if (i == syncedList.size() || !Objects.equals(masterList.get(i), syncedList.get(i))) {
                    syncedList.add(i, masterList.get(i));
                    this.mAdapter.notifyItemInserted(this.mAdapter.getNonNotifItemCount() + i);
                    break;
                }
                i++;
            }
            this.mAdapter.updateRowVisibility();
        }

        private void updateTick() {
            postOneInsert();
            scheduleUpdateTickIfNeeded();
        }

        public void onUiVisible() {
            this.mAdapter.postAllRowUpdates();
        }

        public void onUiInvisible() {
            unScheduleUpdateTick();
        }

        public void onNewRowChange() {
            if (this.mAdapter.getUiState() != 1) {
                return;
            }
            if (this.mIsIdle) {
                this.mAdapter.postAllRowUpdates();
                return;
            }
            postDeletesAndSubstitutes();
            scheduleUpdateTickIfNeeded();
        }

        public void onIdleStateChange(boolean isIdle) {
            this.mIsIdle = isIdle;
            if (isIdle) {
                this.mAdapter.postAllRowUpdates();
                unScheduleUpdateTick();
            }
        }
    }

    protected abstract boolean isPartnerClient();

    protected abstract void onRecommendationDismissed(TvRecommendation tvRecommendation);

    NotificationsViewAdapter(Context context, long minUpdateTime, long maxUpdateTime) {
        super(context);
        if (!isPartnerClient()) {
            this.mPrioritizeRowUpdateState = new PrioritizeRowUpdateState(this, minUpdateTime, maxUpdateTime);
        }
    }

    protected void onClearRecommendations(int reason) {
        if (this.mHomeScreenMessaging != null) {
            this.mHomeScreenMessaging.onClearRecommendations(reason);
        }
    }

    public final int getItemCount() {
        return this.mSyncedList.size() + getNonNotifItemCount();
    }

    public void onIdleStateChange(boolean isIdle) {
        if (!isPartnerClient()) {
            this.mPrioritizeRowUpdateState.onIdleStateChange(isIdle);
        }
    }

    public void onVisibilityChange(boolean isVisible) {
    }

    protected final TvRecommendation getRecommendation(int position) {
        return this.mSyncedList.get(position - getNonNotifItemCount());
    }

    private void purgeDismissedNotifications() {
        for (TvRecommendation rec : this.mRecommendationToBeRemoved) {
            removeRecommendation(rec);
            onRecommendationDismissed(rec);
        }
        this.mRecommendationToBeRemoved.clear();
    }

    protected final void dismissNotification(TvRecommendation rec) {
        this.mRecommendationToBeRemoved.add(rec);
    }

    public final void setNotificationRowViewFlipper(HomeScreenMessaging homeScreenMessaging) {
        this.mHomeScreenMessaging = Preconditions.checkNotNull(homeScreenMessaging);
    }

    private void updateRowVisibility() {
        if (this.mHomeScreenMessaging != null) {
            this.mHomeScreenMessaging.recommendationsUpdated(this.mSyncedList.size() > 0);
        }
    }

    public final void onUiVisible() {
        this.mUiState = 1;
        if (isPartnerClient()) {
            postAllRowUpdates();
        } else {
            this.mPrioritizeRowUpdateState.onUiVisible();
        }
    }

    public final void onUiInvisible() {
        this.mUiState = 2;
        purgeDismissedNotifications();
        if (!isPartnerClient()) {
            this.mPrioritizeRowUpdateState.onUiInvisible();
        }
    }

    public boolean isUiVisible() {
        return this.mUiState == 1;
    }

    int getUiState() {
        return this.mUiState;
    }

    void notifyChangesSinceSnapshot(final List<TvRecommendation> snapshot) {
        DiffResult diffResult = DiffUtil.calculateDiff(new Callback() {
            public int getOldListSize() {
                return snapshot.size();
            }

            public int getNewListSize() {
                return NotificationsViewAdapter.this.mSyncedList.size();
            }

            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return NotificationUtils.equals(snapshot.get(oldItemPosition), NotificationsViewAdapter.this.mSyncedList.get(newItemPosition));
            }

            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                return Objects.equals(snapshot.get(oldItemPosition), NotificationsViewAdapter.this.mSyncedList.get(newItemPosition));
            }
        });
        final int offset = getNonNotifItemCount();
        diffResult.dispatchUpdatesTo(new ListUpdateCallback() {
            public void onInserted(int position, int count) {
                NotificationsViewAdapter.this.notifyItemRangeInserted(offset + position, count);
            }

            public void onRemoved(int position, int count) {
                NotificationsViewAdapter.this.notifyItemRangeRemoved(offset + position, count);
            }

            public void onMoved(int fromPosition, int toPosition) {
                NotificationsViewAdapter.this.notifyItemMoved(offset + fromPosition, toPosition);
            }

            public void onChanged(int position, int count, Object payload) {
                NotificationsViewAdapter.this.notifyItemRangeChanged(offset + position, count, payload);
            }
        });
    }

    void postAllRowUpdates() {
        List<TvRecommendation> snapshot = new ArrayList<>(this.mSyncedList);
        this.mSyncedList.clear();
        this.mSyncedList.addAll(this.mMasterList);
        notifyChangesSinceSnapshot(snapshot);
        updateRowVisibility();
    }

    private void masterListHasChanged() {
        if (isPartnerClient()) {
            switch (this.mUiState) {
                case 0:
                    postAllRowUpdates();
                    return;
                case 1:
                    postAllRowUpdates();
                    return;
                default:
                    return;
            }
        }
        switch (this.mUiState) {
            case 0:
                postAllRowUpdates();
                return;
            case 1:
            case 2:
                this.mPrioritizeRowUpdateState.onNewRowChange();
                return;
            default:
                return;
        }
    }

    List<TvRecommendation> getSyncedList() {
        return this.mSyncedList;
    }

    List<TvRecommendation> getMasterList() {
        return this.mMasterList;
    }

    protected final void onRecommendationsUpdate() {
        masterListHasChanged();
    }

    private int indexOfMasterRecommendation(TvRecommendation rec) {
        for (int i = 0; i < this.mMasterList.size(); i++) {
            if (NotificationUtils.equals(rec, this.mMasterList.get(i))) {
                return i;
            }
        }
        return -1;
    }

    private void removeRecommendation(TvRecommendation rec) {
        int index = indexOfMasterRecommendation(rec);
        if (index != -1) {
            this.mMasterList.remove(index);
            masterListHasChanged();
            if (Log.isLoggable("NotifViewAdapter", 3)) {
                Log.d("NotifViewAdapter", "Recommendation Removed from position" + (index + getNonNotifItemCount()) + ": " + rec);
            }
        }
    }

    protected int getNonNotifItemCount() {
        return 0;
    }

    protected void notifyNonNotifItemChanged(int position) {
        super.notifyItemChanged(position);
    }

    protected void notifyNonNotifItemRemoved(int position) {
        super.notifyItemRemoved(position);
    }

    protected void notifyNonNotifItemInserted(int position) {
        super.notifyItemInserted(position);
    }

    public void dump(String prefix, PrintWriter writer) {
        writer.println(prefix + getClass().getName());
        prefix = prefix + "  ";
        writer.println(prefix + "mMasterList: " + Arrays.toString(this.mMasterList.toArray()));
        writer.println(prefix + "mSyncedList: " + Arrays.toString(this.mSyncedList.toArray()));
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/NotificationsViewAdapter.kt
package com.amazon.tv.leanbacklauncher.notifications

import android.content.Context
import android.os.Handler
import android.util.Log
import androidx.collection.ArraySet
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListUpdateCallback
import androidx.recyclerview.widget.RecyclerView
import com.amazon.tv.leanbacklauncher.MainActivity.IdleListener
import com.amazon.tv.leanbacklauncher.notifications.NotificationUtils.equals
import com.amazon.tv.leanbacklauncher.util.Preconditions
import com.amazon.tv.leanbacklauncher.widget.RowViewAdapter
import com.amazon.tv.tvrecommendations.TvRecommendation
import java.io.PrintWriter
import java.util.*

abstract class NotificationsViewAdapter<VH : RecyclerView.ViewHolder?> internal constructor(
    context: Context?,
    minUpdateTime: Long,
    maxUpdateTime: Long
) : RowViewAdapter<VH>(context), IdleListener {
    private var mHomeScreenMessaging: HomeScreenMessaging? = null
    private val mMasterList = ArrayList<TvRecommendation?>()
    private var mPrioritizeRowUpdateState: PrioritizeRowUpdateState<VH>? = null
    private val mRecommendationToBeRemoved: MutableSet<TvRecommendation?> = ArraySet<TvRecommendation?>()
    private val mSyncedList = ArrayList<TvRecommendation?>()
    var uiState = 0
        private set

    init {
        if (!isPartnerClient) {
            mPrioritizeRowUpdateState =
                PrioritizeRowUpdateState(this, minUpdateTime, maxUpdateTime)
        }
    }

    companion object {
        const val UI_STATE_VISIBLE = 1
    }

    internal class PrioritizeRowUpdateState<VH : RecyclerView.ViewHolder?>(
        private val mAdapter: NotificationsViewAdapter<VH>,
        private val mMinUpdateTime: Long,
        private val mMaxUpdateTime: Long
    ) {
        private val mHandler = Handler()
        private var mIsIdle = false
        private val mStateTickRunnable = Runnable { updateTick() }
        private fun scheduleUpdateTickIfNeeded() {
            val masterList = mAdapter.masterList
            val syncedList: List<TvRecommendation?> = mAdapter.syncedList
            mHandler.removeCallbacks(mStateTickRunnable)
            if (masterList.size - syncedList.size > 0) {
                mHandler.postDelayed(
                    mStateTickRunnable,
                    (mMaxUpdateTime - mMinUpdateTime) / (masterList.size - syncedList.size).toLong() + mMinUpdateTime
                )
            }
        }

        private fun unScheduleUpdateTick() {
            mHandler.removeCallbacks(mStateTickRunnable)
        }

        private fun postDeletesAndSubstitutes() {
            val masterList = mAdapter.masterList
            val syncedList = mAdapter.syncedList
            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize(): Int {
                    return syncedList.size
                }

                override fun getNewListSize(): Int {
                    return masterList.size
                }

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    return equals(syncedList[oldItemPosition], masterList[newItemPosition])
                }

                override fun areContentsTheSame(
                    oldItemPosition: Int,
                    newItemPosition: Int
                ): Boolean {
                    return syncedList[oldItemPosition] == masterList[newItemPosition]
                }

                override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
                    return masterList[newItemPosition]
                }
            }, false)
            val snapshot: List<TvRecommendation?> = ArrayList<TvRecommendation?>(syncedList)
            result.dispatchUpdatesTo(object : ListUpdateCallback {
                override fun onInserted(position: Int, count: Int) {
                    for (i in 0 until count) {
                        syncedList.add(position + i, null)
                    }
                }

                override fun onRemoved(position: Int, count: Int) {
                    for (i in 0 until count) {
                        syncedList.removeAt(position)
                    }
                }

                override fun onMoved(fromPosition: Int, toPosition: Int) {
                    throw IllegalStateException("Should not receive move events")
                }

                override fun onChanged(position: Int, count: Int, payload: Any?) {
                    for (i in 0 until count) {
                        syncedList[position + i] = payload as TvRecommendation?
                    }
                }
            })
            val iterator = syncedList.iterator()
            while (iterator.hasNext()) {
                if (iterator.next() == null) {
                    iterator.remove()
                }
            }
            mAdapter.notifyChangesSinceSnapshot(snapshot)
            mAdapter.updateRowVisibility()
        }

        private fun postOneInsert() {
            val masterList = mAdapter.masterList
            val syncedList: MutableList<TvRecommendation?> = mAdapter.syncedList
            var i = 0
            while (i < masterList.size && i <= syncedList.size) {
                if (i == syncedList.size || masterList[i] != syncedList[i]) {
                    syncedList.add(i, masterList[i])
                    mAdapter.notifyItemInserted(mAdapter.nonNotifItemCount + i)
                    break
                }
                i++
            }
            mAdapter.updateRowVisibility()
        }

        private fun updateTick() {
            postOneInsert()
            scheduleUpdateTickIfNeeded()
        }

        fun onUiVisible() {
            mAdapter.postAllRowUpdates()
        }

        fun onUiInvisible() {
            unScheduleUpdateTick()
        }

        fun onNewRowChange() {
            if (mAdapter.uiState != 1) {
                return
            }
            if (mIsIdle) {
                mAdapter.postAllRowUpdates()
                return
            }
            postDeletesAndSubstitutes()
            scheduleUpdateTickIfNeeded()
        }

        fun onIdleStateChange(isIdle: Boolean) {
            mIsIdle = isIdle
            if (isIdle) {
                mAdapter.postAllRowUpdates()
                unScheduleUpdateTick()
            }
        }
    }

    protected abstract val isPartnerClient: Boolean
    protected abstract fun onRecommendationDismissed(tvRecommendation: TvRecommendation?)
    protected fun onClearRecommendations(reason: Int) {
        if (mHomeScreenMessaging != null) {
            mHomeScreenMessaging!!.onClearRecommendations(reason)
        }
    }

    override fun getItemCount(): Int {
        return mSyncedList.size + nonNotifItemCount
    }

    override fun onIdleStateChange(isIdle: Boolean) {
        if (!isPartnerClient) {
            mPrioritizeRowUpdateState?.onIdleStateChange(isIdle)
        }
    }

    override fun onVisibilityChange(isVisible: Boolean) {}
    protected fun getRecommendation(position: Int): TvRecommendation? {
        return mSyncedList[position - nonNotifItemCount]
    }

    private fun purgeDismissedNotifications() {
        for (rec in mRecommendationToBeRemoved) {
            removeRecommendation(rec)
            onRecommendationDismissed(rec)
        }
        mRecommendationToBeRemoved.clear()
    }

    protected fun dismissNotification(rec: TvRecommendation?) {
        mRecommendationToBeRemoved.add(rec)
    }

    fun setNotificationRowViewFlipper(homeScreenMessaging: HomeScreenMessaging) {
        mHomeScreenMessaging = Preconditions.checkNotNull(homeScreenMessaging)
    }

    private fun updateRowVisibility() {
        if (mHomeScreenMessaging != null) {
            mHomeScreenMessaging!!.recommendationsUpdated(mSyncedList.size > 0)
        }
    }

    fun onUiVisible() {
        uiState = 1
        if (isPartnerClient) {
            postAllRowUpdates()
        } else {
            mPrioritizeRowUpdateState?.onUiVisible()
        }
    }

    fun onUiInvisible() {
        uiState = 2
        purgeDismissedNotifications()
        if (!isPartnerClient) {
            mPrioritizeRowUpdateState?.onUiInvisible()
        }
    }

    val isUiVisible: Boolean
        get() = uiState == 1

    fun notifyChangesSinceSnapshot(snapshot: List<TvRecommendation?>) {
        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return snapshot.size
            }

            override fun getNewListSize(): Int {
                return mSyncedList.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return equals(snapshot[oldItemPosition], mSyncedList[newItemPosition])
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return snapshot[oldItemPosition] == mSyncedList[newItemPosition]
            }
        })
        val offset = nonNotifItemCount
        diffResult.dispatchUpdatesTo(object : ListUpdateCallback {
            override fun onInserted(position: Int, count: Int) {
                notifyItemRangeInserted(offset + position, count)
            }

            override fun onRemoved(position: Int, count: Int) {
                notifyItemRangeRemoved(offset + position, count)
            }

            override fun onMoved(fromPosition: Int, toPosition: Int) {
                notifyItemMoved(offset + fromPosition, toPosition)
            }

            override fun onChanged(position: Int, count: Int, payload: Any?) {
                this@NotificationsViewAdapter.notifyItemRangeChanged(
                    offset + position,
                    count,
                    payload
                )
            }
        })
    }

    fun postAllRowUpdates() {
        val snapshot: List<TvRecommendation?> = ArrayList(
            mSyncedList
        )
        mSyncedList.clear()
        mSyncedList.addAll(mMasterList)
        notifyChangesSinceSnapshot(snapshot)
        updateRowVisibility()
    }

    private fun masterListHasChanged() {
        if (isPartnerClient) {
            when (uiState) {
                0 -> {
                    postAllRowUpdates()
                    return
                }
                1 -> {
                    postAllRowUpdates()
                    return
                }
                else -> return
            }
        }
        when (uiState) {
            0 -> {
                postAllRowUpdates()
                return
            }
            1, 2 -> {
                mPrioritizeRowUpdateState?.onNewRowChange()
                return
            }
            else -> return
        }
    }

    val syncedList: MutableList<TvRecommendation?>
        get() = mSyncedList
    val masterList: MutableList<TvRecommendation?>
        get() = mMasterList

    protected fun onRecommendationsUpdate() {
        masterListHasChanged()
    }

    private fun indexOfMasterRecommendation(rec: TvRecommendation?): Int {
        for (i in mMasterList.indices) {
            if (equals(rec, mMasterList[i])) {
                return i
            }
        }
        return -1
    }

    private fun removeRecommendation(rec: TvRecommendation?) {
        val index = indexOfMasterRecommendation(rec)
        if (index != -1) {
            mMasterList.removeAt(index)
            masterListHasChanged()
            if (Log.isLoggable("NotifViewAdapter", Log.DEBUG)) {
                Log.d(
                    "NotifViewAdapter",
                    "Recommendation Removed from position" + (index + nonNotifItemCount) + ": " + rec
                )
            }
        }
    }

    protected open val nonNotifItemCount: Int
        get() = 0

    protected fun notifyNonNotifItemChanged(position: Int) {
        super.notifyItemChanged(position)
    }

    protected fun notifyNonNotifItemRemoved(position: Int) {
        super.notifyItemRemoved(position)
    }

    protected fun notifyNonNotifItemInserted(position: Int) {
        super.notifyItemInserted(position)
    }

    fun dump(prefix: String, writer: PrintWriter) {
        var prefix = prefix
        writer.println(prefix + javaClass.name)
        prefix = "$prefix  "
        writer.println(prefix + "mMasterList: " + mMasterList.toTypedArray().contentToString())
        writer.println(prefix + "mSyncedList: " + mSyncedList.toTypedArray().contentToString())
    }

}